### ISSUE TITLE
Weird saving issue for nested m2a in m2m

### ISSUE THREAD

**Issue by MiniDigger** [directus/directus#5681]
rc64

I have a really weird error that came out of nowhere.
I have a m2m, with a nested m2a relation, (page m2m paragraph m2a textelement, picturelement).
When I go to edit a text element, via page thru paragraph in the app, it doesn't seem to save the changes.
The PATCH request the app sends
```
{
   "paragraph":[
      {
         "id":16,
         "paragraph_id":{
            "id":4,
            "elements":[
               {
                  "item":{
                     "0":"1",
                     "1":"0",
                     "text":"<h1>New Content</h1>"
                  },
                  "id":10,
                  "paragraph_id":{
                     "name":"Old title"
                  }
               }
            ],
            "name":"New title"
         },
         "sort":0
      }
   ]
}
```

the app displays this: 
```
[INVALID_PAYLOAD] Can't update nested record "paragraph_elements.item" without field paragraph_elements.collection" being set
```

when I enter the text element in the app, console logs this:
```
app.785ff34b.js:1 Uncaught (in promise) TypeError: Cannot read property 'startsWith' of undefined
    at d (app.785ff34b.js:1)
    at B.Object.immediate (app.785ff34b.js:1)
    at w (chunk-vendors.eab02622.js:36)
    at Array.<anonymous> (chunk-vendors.eab02622.js:36)
    at Ge (chunk-vendors.eab02622.js:36)
    at a.ze (chunk-vendors.eab02622.js:36)
    at nt (chunk-vendors.eab02622.js:181)
    at a.Tn.e.$emit (chunk-vendors.eab02622.js:181)
    at Bn (chunk-vendors.eab02622.js:181)
    at Zn (chunk-vendors.eab02622.js:181)
```

relevant code being:
```
const t = b.value, a = t.startsWith("directus_") ? `/${t.substring(9)}/${e.relatedPrimaryKey}` : `/items/${t}/${encodeURIComponent(e.relatedPrimaryKey)}`;
```
which should be this line https://github.com/directus/directus/blob/main/app/src/views/private/components/drawer-item/drawer-item.vue#L306

check seems to return null because item.value is null here
https://github.com/directus/directus/blob/main/app/src/views/private/components/drawer-item/drawer-item.vue#L345

really not sure how to further debug this :/

**Comment by rijkvanzanten** [comment]
Hmm, the API error seems to be accurate, as the payload does in fact not include that collection field for the element, so this must be an app bug.

Another funky thing is that `0: 1` situation:

```
                  "item":{
                     "0":"1",
                     "1":"0",
                     "text":"<h1>New Content</h1>"
                  },
```

**Comment by MiniDigger** [comment]
of course I can't repro this in my dev env, woooo.
I tried comparing the tables, especially directus relations, but can't seem to find a difference. if you got any idea which fields in which table could have been corrupted, that would be appreciated. 
I can't somehow tell directus to reset those relations, right? if I delete the relation, that would delete the existing data.

**Comment by rijkvanzanten** [comment]
Could you check the `one_collection_field` value for directus_relations? That should point to the field that holds the "collection" name

**Comment by MiniDigger** [comment]
absatz = paragraph
elemente = elements
seite = page

| id | many\_collection | many\_field | many\_primary | one\_collection | one\_field | one\_primary | one\_collection\_field | one\_allowed\_collections | junction\_field | sort\_field |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| 40 | absatz\_elemente | absatz\_id | id | absatz | elemente | id | NULL | NULL | item | NULL |
| 41 | absatz\_elemente | item | id | NULL | NULL | NULL | collection | textelement,bildelement | absatz\_id | NULL |
| 43 | seite\_absatz | absatz\_id | id | absatz | NULL | id | NULL | NULL | seite\_id | NULL |
| 44 | seite\_absatz | seite\_id | id | seite | absatz | id | NULL | NULL | absatz\_id | sort |
| 45 | seite\_absatz | absatz\_id | id | absatz | NULL | id | NULL | NULL | seite\_id | NULL |

**Comment by MiniDigger** [comment]
New week, new look. Managed to export my dev DB to my private PC where I can get directus to run from source.
These are the vue warns and unobfuscated errors, after opening the text element in the paragraph in the site (10 being the ID of the text element):
```
vue.runtime.esm.js?2b0e:619 [Vue warn]: Invalid prop: type check failed for prop "edits". Expected Object, got String with value "10".

found in

---> <VForm> at src/components/v-form/v-form.vue
       <TransitionGroup>
         <TransitionDialog> at src/components/transition/dialog/transition-dialog.vue
           <PortalTarget>
             <App> at src/app.vue
               <Root>
vue.runtime.esm.js?2b0e:619 [Vue warn]: Invalid prop: type check failed for prop "edits". Expected Object, got String with value "10".

found in

---> <VForm> at src/components/v-form/v-form.vue
       <TransitionGroup>
         <TransitionDialog> at src/components/transition/dialog/transition-dialog.vue
           <PortalTarget>
             <App> at src/app.vue
               <Root>

drawer-item.vue?fbd1:306 Uncaught (in promise) TypeError: Cannot read property 'startsWith' of undefined
    at fetchRelatedItem (drawer-item.vue?fbd1:306)
    at Object.immediate (drawer-item.vue?fbd1:265)
    at applyCb (vue-composition-api.umd.js?e4fd:1213)
    at Array.eval (vue-composition-api.umd.js?e4fd:1156)
    at flushQueue (vue-composition-api.umd.js?e4fd:1068)
    at VueComponent.flushPostQueue (vue-composition-api.umd.js?e4fd:1027)
    at invokeWithErrorHandling (vue.runtime.esm.js?2b0e:1854)
    at VueComponent.Vue.$emit (vue.runtime.esm.js?2b0e:3888)
    at callHook (vue.runtime.esm.js?2b0e:4223)
    at callUpdatedHooks (vue.runtime.esm.js?2b0e:4351)
vue.runtime.esm.js?2b0e:619 [Vue warn]: Invalid prop: type check failed for prop "edits". Expected Object, got String with value "10".

found in

---> <VForm> at src/components/v-form/v-form.vue
       <TransitionGroup>
         <TransitionDialog> at src/components/transition/dialog/transition-dialog.vue
           <PortalTarget>
             <App> at src/app.vue
               <Root>
vue.runtime.esm.js?2b0e:619 [Vue warn]: Invalid prop: type check failed for prop "edits". Expected Object, got String with value "10".

found in

---> <VForm> at src/components/v-form/v-form.vue
       <TransitionGroup>
         <TransitionDialog> at src/components/transition/dialog/transition-dialog.vue
           <PortalTarget>
             <App> at src/app.vue
               <Root>
```

That leads me to believe, that something in the UI isn't fetching the related field properly.

When I enter into the paragraph, the following requests are triggered:
http://localhost:8081/items/paragraph/4 -> `{"data":{"id":4,"name":"Test","elementes":[10]}}`
http://localhost:8081/items/paragraph/4?fields[]=name&fields[]=elementes-> `{"data":{"name":"Test","elementes":[10]}}`
http://localhost:8081/items/paragraph_elementes?filter=%7B%22id%22:%7B%22_in%22:[10]%7D%7D&fields[]=id&fields[]=item&fields[]=collection&fields[]=null -> `{"data":[{"id":10,"item":"10","collection":"textelement"}]}`
http://localhost:8081/items/textelement?filter=%7B%22id%22:%7B%22_in%22:[%2210%22]%7D%7D&fields[]=type&fields[]=text&fields[]=id -> `{"data":[{"type":"body","text":"<h1>Test2</h1>","id":10}]}`

I have to admit that I never used the rest api before, so I can't tell if there is something obviously wrong (altho that `&fields[]=null` looks strange?)


on saving, I no longer get an error, but its also not saving the inner stuff, its just passing the ID of the paragraph:
```
{"data":{"id":2,"status":"published","sort":null,"date_created":"2020-11-16T12:05:09+01:00","date_updated":"2021-05-17T16:30:40+02:00","titel":"Test Seite2","url":"testseite","template":1,"hide_in_nav":false,"name":null,"sad":[24,25,31],"paragraph":[16]}}
```

this is how the state of the v-form of the drawer for the text element looks like: https://i.imgur.com/xQBU2Pp.png
this the drawer for paragraph: https://i.imgur.com/pWPWSBU.png

**Comment by rijkvanzanten** [comment]
Thanks for that in-depth look @MiniDigger. 

> altho that `&fields[]=null` looks strange?

That definitely looks like something that shouldn't be

**Comment by MiniDigger** [comment]
mmmh, that null is only a missing/empty sort field, that doesn't affect anything as far as I can see.

**Comment by jpetryk** [comment]
I am not sure if this is the same/ related issue or not. But for me there's a really significant issue with content overwriting content within other relations unless they are saved at the highest level after each change is made. This can be demonstrated most easily via video:

https://www.loom.com/share/86c6a9ee913b4361ae25775989532248

I've demonstrated this with files. But it happens with other types of content as well if I follow the same pattern during an update.

Let me know if this is related to this issue, or if I should create a new issue.

**Comment by MiniDigger** [comment]
heh and I thought I had complicated relations :D
from the first look, this doesn't seem related tho, have you checked the browser console? 
my issue will print `TypeError: Cannot read property 'startsWith' of undefined` to the console and is most likely caused by a corrupted DB for the relation (I just need to figure out how...).

if its not the same, best to open a new issue (also, if possible, search for the patch request that saves the collection and add the request body of that to your issue, should be this request when you save https://i.imgur.com/pXEGSkM.png)

**Comment by MiniDigger** [comment]
@rijkvanzanten you will not believe this. after ~3 hours of more or less productive debugging (aka throwing in console logs and comparing a clean new relation chain to my broken one) I finally narrowed down the difference:

my display for the page -> paragraph m2m didn't have a display template set. 

that was enough to "corrupt" the internal state of the app by fetching too much data (list-m2m, adjustFieldsForDisplay) so that the m2a doesn't do what it should do. at least thats what I think is happening. 

so solution for me was adding a display template. I am not sure what to do with this issue now. if a display template is mandatory, it should be marked as required in the UI, but I don't think its mandatory.
so instead somebody (and thats not be, I have no clue how this stuff works, lol) should look why leaving out the display template breaks the nested m2a. I can build an repro cases if needed.

@jpetryk maybe its worth to you to check that all your relations have a display template set, maybe that "solves" your issue too.

**Comment by benhaynes** [comment]
That's odd, without a Display Template, the Primary Key should just be used. 🤔

**Comment by MiniDigger** [comment]
it seems like without a display template, the app fetches the full entry with all nested data, and that confuses some other part of the app. (you can tell I am mostly guessing, I still don't understand much about the data flow in the app)

**Comment by rijkvanzanten** [comment]
> That's odd, without a Display Template, the Primary Key should just be used.

Default to the collection display template, and then the primary key, but yeah

**Comment by jpetryk** [comment]
@MiniDigger thanks for the suggestion to check the display template. In my case, it already has one. I can also verify that when I change one item, save, go into the other item (where it displays the change to the field from the first one), it comes back from the API properly. It's like it believes I'm making the change from the first one in all of them though.

To demonstrate, the attached screenshot shows the API call for the item with ID 22 being correct. Yet the interface is displaying the change I made to ID 21. It actually doesn't matter if, after making the change to the first one, I save it or just click the X. The damage is done. All will be displaying the tentative change if I go in them. Again, if I don't go into any others and save it all the way through to the top level, it will only make the expected change. The oddities occur only after I click on any others prior to saving.

The workaround is having to make changes one at a time, being careful to not click anything other than the one I am changing, saving at the top level each time. Obviously not ideal.

This is a different issue because setting the display template doesn't change anything, and I _think_ my issue actually affects M2O (and perhaps not others). As a result, I'll peel this off into a new issue.

![Screenshot 2021-05-18 22 32 04](https://user-images.githubusercontent.com/5734619/118757699-d4027480-b82a-11eb-8a9d-749e4205b860.png)

**Comment by MiniDigger** [comment]
After talking to nitwel and consistently being able to repro this on rc64, I can't repro this on rc92, so I will close this.
Turns out, 30 releases and 6 months of work make quite the difference!
